src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+* $Id$
+* Portable Audio I/O Library
+* Ring Buffer utility.
+*
+* Author: Phil Burk, http://www.softsynth.com
+* modified for SMP safety on OS X by Bjorn Roche.
+* also allowed for const where possible.
+* modified for multiple-byte-sized data elements by Sven Fischer
+*
+* This program is distributed with the PortAudio Portable Audio Library.
+* For more information see: http://www.portaudio.com
+* Copyright (c) 1999-2000 Ross Bencina and Phil Burk
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction,
+* including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software,
+* and to permit persons to whom the Software is furnished to do so,
+* subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+* ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+* CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+#include "util/math.h"
+#include "util/sample.h"
+
+using ReadableSlice = mixxx::SampleBuffer::ReadableSlice;
+using WritableSlice = mixxx::SampleBuffer::WritableSlice;
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_firstInputBuffer(true),
+          m_writePos(0),
+          m_buffer(bufferSize) {
+    // Set the ring delay buffer items to 0.
+    m_buffer.fill(0);
+}
+
+void RingDelayBuffer::copy(const ReadableSlice pSourceBuffer,
```

Comment:
Since this is a free function now, we this can go to an an anonymous namespace or if we think we can use it elsewhere, it can be moved sample.cpp. 
This is not a plain copy it is a `copyRing()` or such. 

Does it work if the source sourcePos AND destPos are not 0? I think not. We may either assert that or add the case with three copy calls.

Reply:
The mentioned problems should be fixed, to summarise:
- copy() renamed to copyRing()
- copyRing() is moved into an anonymous namespace and serves as a helper function only

> Does it work if the source sourcePos AND destPos are not 0? I think not. We may either assert that or add the case with three copy calls.

I think I understand, what you mean. IMO for the case that such a huge `numItems` value will be provided, which will be many times greater than the size of the source and destination buffer too, it would be needed much more copies than just three. It cannot occur for the actual usage but may be possible. Based on that multiple copies would not be used now, I would prefer the assert-way solution.

Reply:
I have an other (unused) case in mind for three copies: Copy from a ring to a ring.

Source:
123456789

Destination
123456789

Copy 8 samples 
Read pointer at 5 write pointer at 7
Copy
5...7 to 7...9 (3)
7...9 to 1..3 (3)
1..2 to 4..5

This can't happen if one of the pointers points to the start.

Reply:
Asserting that this not happens works for me.

Reply:
Oh I missed that you did it already.
Thank you.